plugins/catalog-backend/src/ingestion/processors/microsoftGraph/read.ts

```diff
@@ -0,0 +1,278 @@
+  })) {
+    if (!group.id || !group.displayName) {
+      continue;
+    }
+
+    const name = normalizeEntityName(group.mailNickname || group.displayName);
+    const entity: GroupEntity = {
+      apiVersion: 'backstage.io/v1alpha1',
+      kind: 'Group',
+      metadata: {
+        name: name,
+        description: group.displayName,
+        annotations: {
+          [MICROSOFT_GRAPH_GROUP_ID_ANNOTATION]: group.id,
+        },
+      },
+      spec: {
+        type: 'team',
+        // TODO: We could include a group email and picture
+        ancestors: [],
+        children: [],
+        descendants: [],
+      },
+    };
+
+    const groupMembers = new Set<string>();
+
+    // Download the members in parallel, otherwise it can take quite some time
+    const loadGroupMembers = limiter(async () => {
+      for await (const member of client.getGroupMembers(group.id!)) {
+        if (
+          !member.id ||
+          !(
+            member['@odata.type'] === '#microsoft.graph.user' ||
+            member['@odata.type'] === '#microsoft.graph.group'
+          )
+        ) {
+          continue;
+        }
+
+        groupMembers.add(member.id);
+      }
+    });
+
+    groupMemberPromises.push(loadGroupMembers);
+    groupMember.set(group.id, groupMembers);
+    groups.push(entity);
+  }
+
+  // Wait for all group members to be loaded
+  await Promise.all(groupMemberPromises);
+
+  return {
+    groups,
+    rootGroup,
+    groupMember,
+  };
+}
+
+export function resolveRelations(
+  rootGroup: GroupEntity | undefined,
+  groups: GroupEntity[],
+  users: UserEntity[],
+  groupMember: Map<string, Set<string>>,
+) {
+  // Build reference lookup tables, we reference them by the id the the graph
+  const userMap: Map<string, UserEntity> = new Map(); // by uuid
+  const groupMap: Map<string, GroupEntity> = new Map(); // by uuid
+
+  for (const user of users) {
+    userMap.set(
+      user.metadata.annotations![MICROSOFT_GRAPH_USER_ID_ANNOTATION],
+      user,
+    );
+  }
+  for (const group of groups) {
+    if (group.metadata.annotations![MICROSOFT_GRAPH_GROUP_ID_ANNOTATION]) {
+      groupMap.set(
+        group.metadata.annotations![MICROSOFT_GRAPH_GROUP_ID_ANNOTATION],
+        group,
+      );
+    }
+    if (group.metadata.annotations![MICROSOFT_GRAPH_TENANT_ID_ANNOTATION]) {
+      groupMap.set(
+        group.metadata.annotations![MICROSOFT_GRAPH_TENANT_ID_ANNOTATION],
+        group,
+      );
+    }
+  }
+
+  // Make sure that every user is at least part of the rootGroup
+  if (rootGroup) {
+    const groupMembers = groupMember.get(
+      rootGroup.metadata.annotations![MICROSOFT_GRAPH_TENANT_ID_ANNOTATION],
+    );
+
+    users.forEach(u =>
+      groupMembers?.add(
+        u.metadata.annotations![MICROSOFT_GRAPH_USER_ID_ANNOTATION],
+      ),
+    );
+  }
+
+  // TODO: For every group, add the group members (users/groups might be missing!)
+
+  // TODO: Also added to partent to every member
+
+  buildOrgHierarchy(groups);
+}
```

Comment:
This is the central part that is missing right now. I have:
* The users
* The groups
* The root group
* A map-set with the members of each group

I know wonder how I should combine these to the hierarchy:
* Should every user be at least part of the root group? Otherwise members without a group aren't in the hierarchy at all? We have a lot of these users, but that might be a problem in our data too...
* Which relationships do I need to write in the processor and which are added later on? Looks like `buildOrgHierarchy` only set groups, do I have to set `memberOf` in each user? Is it only required to set direct relationships and do I have to add it in both ways, or is one direction sufficient?
* Should I write them in the entity, or is there a way to expose relations directly?

@freben can you give some input here?

Afterwards I only have to consider the tests for this function.

Reply:
Hi! Written while running toward the dinner table: We said just today that we really should remove the `ancestors` and `descendants` concepts entirely, like, right now. That would remove a small part of the complexity. I'll look closer at your ask later or tomorrow

Reply:
OK, back again. @Rugvip may want to proof read what I write here to see if it aligns with what we've said.

## Missing relations

Users do not have to be part of a group. Free floating users and groups are fine too, from a model perspective, but of course it will have consequences. If a user is actually part of some group but the model isn't aware of it, then that user won't be seen as the owner of components mapped to that group. This affects their Backstage experience aversely.

Org hierarchies not being perfect (some missing parent/child relationships between groups) is fine too, but again, runs the risk of missing transitive ownerships (component owned by A, which is a subgroup of B but the model doesn't know).

A cool thing about Backstage at Spotify, is that it has incentivised people into keeping the org hierarchy clean and properly mapped out - it gives them great benefits to do so and they notice when things are off.

That being said, as you have seen, we haven't yet finalised the part where the owner being a group, gets wired up to a user being part of the group, in the open source Backstage. Only ownership matching the user itself is currently supported. That's of course going to change soon.

## Emitting relations

Relations are meant to be emitted during processing. See the [OwnerRelationProcessor](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/OwnerRelationProcessor.ts) for an example. You'll notice that the source and target are refs (kind/namespace/name triplets), not uids, and do NOT have to map to an existing or even future entity. Dangling references are accepted.

You MAY emit relations that aren't represented in the `spec` at all. You SHOULD always emit both sides of a relation when you do, because you don't know from which angle they will be consumed.

NOTE: A point to consider here is that the relations can come from either end. Some data sources may have data with `children` filled in but no `parent`, other sources may have the inverse and only supply `parent`. Both should in the end result in the exact same set of relations. Same with groups either listing their immediate members, or users listing the groups they are immediate members of.

## How the frontend will use relations for login and ownership

So with that out of the way: The frontend should end up having the following needs.

When you log in, you will be mapped to a backstage identity (a user in the catalog). The identity provider will probably be responsible for doing so, e.g. by asking for `kind=user,spec.profile.email=<email>` if the provider presents an email based identity.

In the response for that entity, there should be a number of `memberOf group:default/x` relations. The frontend should hold on to that list of groups. Let's say we got `x`, `y`, and `z`. Maybe `x` is a `childOf` `y` and `y` is a `childOf` `z`; at that point we do not really care, but it does imply that users should have all transitive `memberOf` relations. Otherwise they won't be able to transitively own things.

When looking at the catalog page and listing the entities that the user owns, we'll ask the backend for `filter=relations.ownedBy=x&filter=relations.ownedBy=y&filter=relations.ownedBy=z`. There's an OR between multiple `filter`s. The relations filter syntax is TBD though. This should give you the list of entities that's owned by either one of `x`, `y`, and `z`. This seems to imply that the `ownedBy` relation doesn't strictly have to be generated transitively, e.g. if we have `spec.owner: y`, we don't have to generate `ownedBy z` as well if `z` is a parent of `y` - at least not for the purposes of this exercise.

## So, what to do now

The end result needs to be a lot of relations. We could either make it so that we generate relations directly from a bunch of groups and users, or we do what the old code did to fill in `spec` data in the users and groups which later downstream get translated into relations by a generic post-processor. Filling in at least SOME of the transitive relations (as per above) will be important. That will probably be easiest to do immediately after ingestion, when we are holding all of the groups and users at the same time, unless we invent a future processing step that can do "global"/"inter-entity" analysis.

Reply:
First: VERY valuable feedback. Thank you! I will have a deeper look tomorrow.

Reply:
**Should every user be at least part of the root group?**

I think I will not do that. So we have a reason to put every user into useful groups. With Backstage the org hierarchy will be much more visible, not just to our admins and then we have a reason to improve it. 

But I will add another behavior: Every group that has no parent group will be part of the root group. I need to implement this myself as the datasource itself doesn't expose the relations for tenants explicitly (only implicitly by the authentication).

**Should I emit the relationships directly or should I include in the entity spec?**

Not sure about your answer here. I will try to emit them directly, if that is too complicated I will include them in the spec for now (I will have a look at `OwnerRelationProcessor`).

** Should I emit transitive relationships?**

If I understand it correctly, having the relationships between groups is sufficient. So I will go with `buildOrgHierarchy`.

**Should I always emit a relationship in both ways?**

> You SHOULD always emit both sides of a relation when you do, because you don't know from which angle they will be consumed.

That is a yes 😉

Reply:
> **Should every user be at least part of the root group?**
> 
> I think I will not do that. So we have a reason to put every user into useful groups. With Backstage the org hierarchy will be much more visible, not just to our admins and then we have a reason to improve it.

It's an interesting point to talk about "useful" groups, and what constitutes usefulness, and whether usefulness should be part of the filtering criteria.

Ultimately maybe this will be up to the org at hand, and what you want to use the information for. The root group is not much more than a "virtual" kind of placeholder in most orgs, I would think; It wouldn't hopefully OWN anything for example. At Spotify, we also have a mashup of non-org-tree groups as part of the groups list - such as pure email groups (that aren't tied to anything at all, except perhaps being parts of an even larger email group), virtual groups that aren't SUPPOSED to own things per our rules but still sometimes do, marker groups that are used exclusively for access control, etc. So what does one want the catalog to represent? We've found that adding these marker groups sometimes is detrimental to performance and clarity (pretty much every freaking person has access to some things, and there are many things, and it just messes up the user interface to list them alongside the relevant ones - for some definition of relevant).

Hot take: It's fine if the Backstage catalog isn't a full mirror of the authoritative source, but rather only contains the subsets that make sense for its own successful operation.

We also have a groups management interface in our internal Backstage, made by the team that owns the authoritative source. THAT plugin talks directly to the authoritative source for all of its needs, so it can show ALL group memberships and their intricate details etc. If you make modifications, all it does with the catalog is to tell it to go refresh itself from the authoritative source. I think that's probably the right balance here too.

> But I will add another behavior: Every group that has no parent group will be part of the root group. I need to implement this myself as the datasource itself doesn't expose the relations for tenants explicitly (only implicitly by the authentication).

Does this solve some particular problem you were running into? Even if you choose to not mirror everything, does it not pose a risk to also "lie" in the data (i.e. not just plain mirror)?

> **Should I emit the relationships directly or should I include in the entity spec?**
> 
> Not sure about your answer here. I will try to emit them directly, if that is too complicated I will include them in the spec for now (I will have a look at `OwnerRelationProcessor`).
>
> **Should I emit transitive relationships?**
> 
> If I understand it correctly, having the relationships between groups is sufficient. So I will go with `buildOrgHierarchy`.

Yeah immediate (nearest) relationships between groups is probably enough for now. That means that one does NOT need a "global" view of groups to emit `parentOf`/`childOf` for them.

To me, that seems to imply that the best path forward right now is to use `buildOrgHierarchy` to build out the `spec`, and then introduce a new processor class in the core offering that emits relations based on that spec. This makes the processor useful for those that prefer to go via the `spec`, and it'll just do nothing for those users that instead choose to emit in their own reader and leave `spec` empty.

Hot take: I think we could build out the [BuiltinKindsEntityProcessor](https://github.com/backstage/backstage/blob/master/plugins/catalog-backend/src/ingestion/processors/BuiltinKindsEntityProcessor.ts) to get a `postProcessEntity` that emits these (and later, other) relations automatically.

Reply:
> It's an interesting point to talk about "useful" groups, and what constitutes usefulness, and whether usefulness should be part of the filtering criteria.

Yeah I use the `groupFilter` to already filter out groups that are not needed, like security only groups.

> Does this solve some particular problem you were running into? Even if you choose to not mirror everything, does it not pose a risk to also "lie" in the data (i.e. not just plain mirror)?

Not lie, the groups are part of the tenant because I can query them when I'm authenticated as the tenant, so they are all part of the root group.
Not sure why I'm so fixed into wanting to have a root group. I could also just leave the tenant/org out and go without a root group at all? @freben What do you think?

> ...buildOrgHierarchy to build out the spec, and then extend the core offering to emit relations based on that spec. 
Yeah, you are right. If I use `buildOrgHierachy` I should stay with the relations in the spec.

My goal to finish this up later today, thanks for the input.

Reply:
> Not lie, the groups are part of the tenant because I can query them when I'm authenticated as the tenant, so they are all part of the root group.

Alright, I see. So it's more about how the Microsoft Graph API thinks about things.

> Not sure why I'm so fixed into wanting to have a root group. I could also just leave the tenant/org out and go without a root group at all? @freben What do you think?

Eh, I say you already went through the effort of it, so keep it. Having a plugin (or native part of the [User/Group pages](https://github.com/backstage/backstage/issues/2752)) that shows your part of the org hierarchy is kinda nice and the alternative is to have a forest of disjoint trees that's harder to navigate, should you want to.